opensquirrel/merger/general_merger.py

```diff
@@ -79,6 +86,41 @@ def try_name_anonymous_bloch(bsr: BlochSphereRotation) -> BlochSphereRotation:
     return bsr
 
 
+def _flatten_list(list_to_flatten: list[list[Any]]) -> list[Any]:
+    return functools.reduce(operator.iadd, list_to_flatten, [])
+
+
+def merge_barriers(statement_list: list[Statement]) -> list[Statement]:
```

Comment:
I think there is a problem with this code, and it shows in the first test, `initial-barrier`.
If we have something like:

```
barrier q[0]
H q[1]
barrier q[1]
H q[0]
```

We should end up with:

```
H q[1]
barrier q[0]
H q[0]
barrier q[1]
```

Right? Correct me if I'm wrong. Because `H q[0]` can go before `barrier q[1]`.

However, because of the way this code works, we end up with:

```
H q[1]
barrier q[0]
barrier q[1]
H q[0]
```

Why does the code behave like that? Because, even if `barrier q[0]` and `barrier q[1]` are not linked in a beginning, once the code treats `barrier q[1]`, it creates a group of linked barriers, over which "H q[0]" cannot go through.

Reply:
Yes, `H q[0]` can go before `barrier q[1]` but it goes against the strategy of merging the barriers. The most important functionality we should strive for is to make sure that these barrier walls across our circuit are correct (and also optimal as is the case above).

Reply:
At face value, the following result is permitted, given the meaning of the barriers:
```
H q[1]
barrier q[0]
H q[0]
barrier q[1]
```
including optimization this would become:
```
H q[1]
barrier q[0]
barrier q[1]
H q[0]
```
But since both restructurings influence the order of statements and such processing should be contained within a (rudimentary) Scheduling pass, which is not part of OpenSquirrel atm, the resulting expected result should be:
```
barrier q[0]
H q[1]
barrier q[1]
H q[0]
```
(basically no change... it couldn't be easier 😅 :)

Reply:
I understood from earlier talks with Chris, that the barriers should strive to be merged unless there was a more optimal way of doing things. This situation shows that, because there is no "optimized" way of doing ordering the gates, it strives to merge the gates because it is, from my understanding, how quantum systems will understand a barrier wall. 

This definition and test are deprecated anyhow since this merge_barrier call will be modified according to our potnetial future `Scheduling` passes.